### ISSUE TITLE
Caddy does not respond correctly to preflight requests with basic authorization

### ISSUE THREAD

**Issue by matthias-k** [mholt/caddy#1568]
### 1. What version of Caddy are you running (`caddy -version`)?

Caddy 0.9.5 with the cors addon downloaded from the caddy website on debian testing 64bit.

### 2. What are you trying to do?

I am trying to serve an API with caddy using basic authorization. The frontend (using ember.js) access the API with CORS  with preflight requests and the basic authorization. However, caddy responds to the preflight requests with 401 if the request does not contain the authorization header (which you are not allowed to send in the preflight request: [CORS spec](https://www.w3.org/TR/cors/#preflight-request)). If I add the authorization header (which I am actually not allowed to do), then caddy responds correctly. In the following I show a minimal example which does not use a reverse proxy but instead just serves the local directory.

### 3. What is your entire Caddyfile?
```text
localhost:8080 {
    log access.log
    errors errors.log
    basicauth / user pass
    cors
}
```

### 4. How did you run Caddy (give the full command and describe the execution environment)?

I run caddy using `./caddy` in the directoy where the caddy file and an `index.html` are present.

### 5. Please paste any relevant HTTP request(s) here.

This is the correct preflight request:

```
$ curl -v 'http://localhost:8080' -X OPTIONS -H 'access-control-request-method: GET' -H 'origin: http://localhost:4200' -H 'access-control-request-headers: authorization'                                        
* Rebuilt URL to: http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> OPTIONS / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.52.1
> Accept: */*
> access-control-request-method: GET
> origin: http://localhost:4200
> access-control-request-headers: authorization
> 
< HTTP/1.1 401 Unauthorized
< Content-Type: text/plain; charset=utf-8
< Server: Caddy
< Www-Authenticate: Basic realm="Restricted"
< X-Content-Type-Options: nosniff
< Date: Fri, 07 Apr 2017 17:37:42 GMT
< Content-Length: 17
< 
401 Unauthorized
* Curl_http_done: called premature == 0
```

And this is the incorrect preflight request which includes the authorization header:

```
$curl -v 'http://localhost:8080' -X OPTIONS -H 'access-control-request-method: GET' -H 'origin: http://localhost:4200' -H 'access-control-request-headers: authorization'  -H "Authorization: Basic dXNlcjpwYXNz"
* Rebuilt URL to: http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> OPTIONS / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.52.1
> Accept: */*
> access-control-request-method: GET
> origin: http://localhost:4200
> access-control-request-headers: authorization
> Authorization: Basic dXNlcjpwYXNz
> 
< HTTP/1.1 200 OK
< Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE
< Access-Control-Allow-Origin: http://localhost:4200
< Server: Caddy
< Vary: Origin
< Date: Fri, 07 Apr 2017 17:38:47 GMT
< Content-Length: 0
< Content-Type: text/plain; charset=utf-8
< 
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

### 6. What did you expect to see?

As far as I know, the first request (excluding the credentials) should be answered as the second request with 200 and an empty response body.

### 7. What did you see instead (give full error messages and/or log)?

See above: caddy responds with 401

### 8. How can someone who is starting from scratch reproduce the bug as minimally as possible?

See above: Serve a directory with caddy with basic authorization and try to send an `OPTIONS` preflight request with does not include the credentials but indicates their presence via `access-control-request-headers: authorization`. In my actual example, caddy acts as a reverse proxy for a python backend but that doesn't affect the problem.

**Comment by mholt** [comment]
I need time to read the whole issue, but would this be something you know anything about, @captncraig ?

**Comment by matthias-k** [comment]
@mholt @captncraig I first tried without the cors plugin. Without the plugin, caddy responds to the OPTIONS request without the `authorization` header in exactly the same way (401) and to the OPTIONS request including the `authorization` header with a 200 GET response which returns `index.html`. I'm not sure whether that is correct, either.

**Comment by captncraig** [comment]
Sounds like a plugin ordering issue to me. The basicauth runs before the cors, so it denies the preflight before cors gets a change to respond.

**Comment by captncraig** [comment]
So maybe basicauth should ignore OPTIONS requests (questionable to me), or else put cors higher up the stack.

**Comment by matthias-k** [comment]
To me it seems like basic auth should check that the `OPTIONS` request indicates the presence of an `Authorization` header via the `access-control-request-headers` but I haven't figured out what the CORS spec says about this.

Additionally one would need to make sure that to caddy does not return the actual data in this case if the cors plugin is not enabled: if the cors plugin is disabled and I send the credentials in the `Authorization` header of the `OPTIONS` request, then caddy responds with a `GET` response and the actual data.

**Comment by mholt** [comment]
I'd like to keep concerns as separate as possible: if running cors before basicauth is all that needs to happen (and won't break anything), that sounds ideal, rather than having basicauth be CORS-aware. I'd accept a PR for that (but double check the list of directives affected by the re-order and make sure it won't break anything).

**Comment by Rurik19** [comment]
Can I write something like that: 
```
if {method} is OPTIONS
    cors
else
   basicauth
```

**Comment by mholt** [comment]
@Rurik19 No; try asking on our [forums](https://caddy.community) as more people will see your question!

**Comment by Rurik19** [comment]
I solved it with rewrite rule, see https://github.com/keeweb/keeweb/issues/706#issuecomment-320303152

**Comment by mholt** [comment]
@Rurik19 Thanks for following up.

@captncraig What do you think on this? Should cors be reordered? Or is that not correct. (I haven't looked into the nuances/spec of CORS...)

**Comment by Rurik19** [comment]
It somewhere here https://www.w3.org/TR/cors/#resource-preflight-requests

> 7.1.5 Cross-Origin Request with Preflight
> Exclude user credentials.

**Comment by nprail** [comment]
@mholt @captncraig I am having this issue as well. Reordering doesn't seem to help. I have basicauth all the way at the bottom and cors near the top and it still requires authentication when sending a `OPTIONS` request.

**Comment by ramigg** [comment]
Is there any advance with that?
thanks

**Comment by theodesp** [comment]
@nprail The reordering you specify in your Caddyfile is different from the actual reordering that the server applies.

**Comment by noreek** [comment]
Looks like this feature is lost in caddy v2 again? Or maybe there is some extra configuration required to disable basic auth for OPTIONS method?

**Comment by noreek** [comment]
Sorry, found a way to do this:
```
@basic_matcher {
        path /*
        not method OPTIONS
}

basicauth @basic_matcher { ... }
```

**Comment by francislavoie** [comment]
@noreek your `path` matcher there is redundant, because that will always be true. Change it to this instead: `@basic_matcher not method OPTIONS`
